ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java

```diff
@@ -323,12 +352,13 @@ public void onResponse(ReplicaId replicaId, TrackedRequestFinalState trackedRequ
   private class OpTrackerIterator implements Iterator<ReplicaId> {
     @Override
     public boolean hasNext() {
-      return inflightCount < parallelism && replicaIterator.hasNext();
+      return inflightCount < getCurrentParallelism() && replicaIterator.hasNext();
```

Comment:
Do we need to keep different `inflightCount` for diskReplica and cloudReplica separately? If inflightCount already contains request cloudReplica and we compare this number to parallelism of diskReplica might be wrong. Let's say diskReplica's parallelism is 3 and now there are 2 inflight diskReplica and 1 inflight cloudReplica, then `hasNext` may return false. However, the actual number of inflight diskReplica is less than parallelism of disk replica.

Reply:
This is an interesting case. There is a question here about whether the cloud and disk parallelism can be summed or not. In your case the effective parallelism could be 4 since you are allowing both disk and cloud requests to reach full parallelism.

However, I feel that although this increases the chance of success for the disk replicas, it is not completely necessary as a single acknowledgement from a cloud replica could satisfy the operation. The reason I decided to shrink the parallelism instead of summing it was to not waste a single cloud request by also sending requests to disk replicas at the same time. I notice that I also made an assumption that the requests are ordered by datacenter, which sometimes isn't true for the backup and down nodes.

Reply:
I feel ok with current logic. Perhaps we can add debug log to help determine if the request being sent is cloudReplica request or disk one.